Improve logging statements in CES usage and reduce code reuse

yanggangtony opened the pull request:
Signed-off-by: yanggang <gang.yang@daocloud.io>

/kind bug
fix panic when map return nil.
fix: [https://github.com/cilium/cilium/issues/22417](https://github.com/cilium/cilium/issues/22417)

Commit:
operator: Refactor to reduce duplicate code

getCESTrackerOnly() is just a slightly different version of
getCESTracker(), we only need one such function. Remove the other one.

Signed-off-by: yanggang <gang.yang@daocloud.io>
Signed-off-by: Joe Stringer <joe@cilium.io>

nebril commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -634,15 +634,16 @@ func (c *cesMgr) insertCESInWorkQueue(ces *cesTracker, baseDelay time.Duration)
+			ces.cesInsertedAt = t
+			diff = timeSinceCESQueued.Seconds()
+		}
 	}
```
It would be great to get a log message here so that the fail is not completely silent.

Also I just realized that there are multiple usages of `getCESTrackerOnly` in this file, can you fix all of them by either adding nil checks or using `getCESTracker`?

nebril reviewed (changes requested):
Thanks for the PR, I think it's a good catch. Please consider expanding it a bit as explained in the comment.

pchaigno reviewed (dismissed):
Several reviewers have asked for the commit description to be expanded. Adding the stack trace there would probably be good enough.

joestringer commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -333,10 +336,9 @@ func (c *cesMgr) RemoveCEPFromCache(cepName string, baseDelay time.Duration) {
 		var ces *cesTracker
 		if ces, exists = c.desiredCESs.getCESTracker(cesName); !exists {
-			log.Info("Attempted to remove non-existent CES, skipping.")
+			log.Infof("Attempted to remove non-existent cesName :%v , skipping.", cesName)
```
Thanks, this kind of info can help debuggability 👍 however, please use structured logging for log messages rather than using the `log.Infof()` varieties. This helps logging systems to aggregate logs appropriately.
```suggestion
			log.WithFields(logrus.Fields{
			    "CiliumEndpointSlice": cesName,
			}.Info("Attempted to remove non-existent CES, skipping.")
```

joestringer replied:
(Same feedback will apply to the other logging messages)

joestringer commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -408,6 +410,8 @@ func (c *cesMgr) getCEPCountInCES(cesName string) (cnt int) {
 		cnt = len(ces.ces.Endpoints)
 		ces.backendMutex.RUnlock()
+	} else {
+		log.Infof("Attempted to getCEPCountInCES non-existent CES : %v, skipping.", cesName)
```
Hmm, is this an info level log message or debug? I am having a bit of a struggle understanding how a user could interpret this message and take action if they see it in their cluster.

joestringer commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -543,10 +548,13 @@ func (c *cesManagerIdentity) InsertCEPInCache(cep *cilium_v2.CoreCiliumEndpoint,
+				c.addCEPtoCES(cep, ces)
+				return cesName
+			} else {
+				log.Infof("Attempted to InsertCEPInCache non-existent cesName : %v, skipping.", cesName)
```
Is this an error condition? I wonder whether we should be consuming such cases at this sort of level or attempting to forward an error up the stack to hint that the CES cannot be properly implemneted.

joestringer reviewed (commented):
I didn't take too deep of a look but I noticed some aspects of the logging that deviates from the usual approach to logging in Cilium. I'm also not sure I understand whether the new logging situations are expected or unexpected conditions. Typically `log.Info()` should be limited to mostly once-off messages that do not recur frequently. If we're hiding an error condition, then maybe we should be reflecting those errors somewhere but I'm not exactly sure where. Probably someone with better CES and/or k8s knowledge could provide more concrete feedback in this regard. cc @Weil0ng @cilium/sig-k8s

joestringer commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -453,11 +464,10 @@ func (c *cesMgr) clearRemovedCEPs(cesName string, remCEPs map[string]struct{}) {
 	if ces, ok = c.desiredCESs.getCESTracker(cesName); !ok {
 		log.WithFields(logrus.Fields{
-			logfields.CESName: cesName,
+			"CESName": cesName,
```
Why is this switching from the string variable to open-writing the name? The reason we use `logfields.CESName` for cases like this is to ensure that all logs that have a field where the field is the CES name will all have the exact same logging key in the structured logs.

Please switch this and all of the other cases to `logfields.CESName`.

yanggangtony replied:
Thanks for review..
Had updated.
But i still found the log label have manual-define, like 'CEPCount', 'totalCEPCount'..
Is that ok? 
or we should move it to the "pkg/logging/logfields/logfields.go" to define it?

joestringer replied:
Could you define all of the keys in `pkg/logging/logfields/logfields.go`? This will keep them all consistent across the codebase. Thanks.

yanggangtony replied:
yes, updated.

varunmar commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -282,18 +285,22 @@ func (c *cesMgr) getCESCopyFromCache(cesName string) (*cilium_v2.CiliumEndpointS
+		} else {
+			log.WithFields(logrus.Fields{
+				logfields.CESName: cesName,
+			}).Debug("Attempted to InsertCEPInCache non-existent cesName , skipping.")
```
Suggest:
"CEP to CES mapping incorrect - missing CESName"

and add both CEP and CES names to the log fields.

varunmar commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -325,18 +332,17 @@ func (c *cesMgr) InsertCEPInCache(cep *cilium_v2.CoreCiliumEndpoint, ns string)
 		var ces *cesTracker
 		if ces, exists = c.desiredCESs.getCESTracker(cesName); !exists {
-			log.Info("Attempted to remove non-existent CES, skipping.")
+			log.WithFields(logrus.Fields{logfields.CEPName: cepName}).Info("Attempted to remove non-existent CES, skipping.")
```
Add CESName here as well.

varunmar commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -361,10 +367,8 @@ func (c *cesMgr) RemoveCEPFromCache(cepName string, baseDelay time.Duration) {
 	} else {
 		log.WithFields(logrus.Fields{
 			logfields.CEPName: cepName,
```
Add CESName to fields here as well.

varunmar reviewed (changes requested):
Thanks for the change!
This is unexpected, and should not be happening - I think the additional debug logs will come in very useful if/when we can reproduce this problem.

I suggested a couple more lines of logging to make finding the problem easier, thanks for the PR!

varunmar reviewed (approved):
Thanks for the changes. Looks good to me now.

Weil0ng reviewed (approved):
Sorry I was OOO, thank you for spotting the issue and sending in the mitigation! Change LGTM. Just curious, is the cluster under heavy pod churn when you encountered the panic?

dlapcevic reviewed (approved):
Hi, I already fixed the nil access panic bug in `getCESQueueDelayInSeconds()` - https://github.com/cilium/cilium/pull/22884
The explanation of how this happens is in the description.

I was unaware of this PR and the issue, we should’ve closed https://github.com/cilium/cilium/issues/22417 when my PR was merged.

I think there is no way it will try to access nil pointer right now, because all parts of the code that use `getCESTrackerOnly()` have checks for it, either by checking if it’s nil or using `getCESName()` to see if it exists.

`getCESName()` is also reliable because `createCES()` and `insertCEP()` are used together, so if `cepNametoCESName` exists, a `cesTracker` for that CEP will also exist.
Code:
getCESTrackerOnly - https://github.com/cilium/cilium/blob/master/operator/pkg/ciliumendpointslice/ceptocesmap.go#L89
getCESName - https://github.com/cilium/cilium/blob/master/operator/pkg/ciliumendpointslice/ceptocesmap.go#L47
createCES - https://github.com/cilium/cilium/blob/master/operator/pkg/ciliumendpointslice/manager.go#L200
insertCEP - https://github.com/cilium/cilium/blob/master/operator/pkg/ciliumendpointslice/ceptocesmap.go#L34

Anyway, it’s good to remove `getCESTrackerOnly()` to avoid future issues with unsafe access.

LGTM

Commit:
operator: Improve logging for CES interactions

Reuse logfields for common names, add some missing fields from logging
statements, and add additional debugging statements in error conditions.

Signed-off-by: yanggang <gang.yang@daocloud.io>
Signed-off-by: Joe Stringer <joe@cilium.io>
